### DISCUSSION TITLE
Coding convention

### DISCUSSION THREAD

**Discussion by seokho-son** [cloud-barista/cb-tumblebug#689] [General]
# 코딩 컨벤션을 논의 및 가이드를 확정하기 위한 공간입니다.

### snake_case or camelCase 
- camelCase 로 확정 (기존 코드는 업데이트 필요. ref: https://github.com/cloud-barista/cb-tumblebug/pull/684)

### initialisms 

주요 논의 사항
- 축약어 (acronym, abbreviation) 의 대소문자 구성 형태 (ex: id, ID, Id). 
- 특히 camelCase 로 작성되는 경우 camel**C**ase 부터 시작하는 경우가 이슈가 됨. (ex: num + vm + id => numVMID / numVmId / ...). 
- 더 까다로운 경우는 변수의 첫 글자가 축약어로 시작하는 경우임 (ex: vm + id => vmID / vmId / VMID / vmid)
- VMID 로 정하는 경우, golang에서는 첫 글자 대문자로 처리 시에 외부 공유하는 것을 의미하기 때문에 의도하지 않은 오픈이 발생할 수 있음.

[참고]
https://github.com/golang/go/wiki/CodeReviewComments#initialisms

> Initialisms
> Words in names that are initialisms or acronyms (e.g. "URL" or "NATO") have a consistent case. 
> For example, "URL" should appear as "URL" or "url" (as in "urlPony", or "URLPony"), never as "Url". As an example: ServeHTTP not ServeHttp. For identifiers with multiple initialized "words", use for example "xmlHTTPRequest" or "XMLHTTPRequest".
> 
> This rule also applies to "ID" when it is short for "identifier" (which is pretty much all cases when it's not the "id" as in "ego", "superego"), so write "appID" instead of "appId".

https://github.com/golang/lint/blob/c5fb716d6688a859aae56d26d3e6070808df29f7/lint.go#L742
> // commonInitialisms is a set of common initialisms.
> // Only add entries that are highly unlikely to be non-initialisms.
> // For instance, "ID" is fine (Freudian code is rare), but "AND" is not.
> var commonInitialisms = map[string]bool{
> 	"ACL":   true,
> 	"API":   true,
>...
> 	"HTTPS": true,
> 	"ID":    true,
> 	"IP":    true,
> 	"JSON":  true,
>...
> 	"UI":    true,
> 	"UID":   true,
> 	"UUID":  true,
> 	"URI":   true,
> 	"URL":   true,
>...
> 	"XML":   true,
>...

## 2. golang struct JSON Tag

https://github.com/golang/go/wiki/Well-known-struct-tags#background
https://stackoverflow.com/questions/10858787/what-are-the-uses-for-tags-in-go

현재 사용 형태
```
// struct TbVmReq is to get requirements to create a new server instance.
type TbVmReq struct {
	// VM name or VM group name if is (not empty) && (> 0). If it is a group, actual VM name will be generated with -N postfix.
	Name string `json:"name"`

	// if vmGroupSize is (not empty) && (> 0), VM group will be gernetad. VMs will be created accordingly.
	VmGroupSize string `json:"vmGroupSize" example:"3" default:""`

	Label string `json:"label"`

	Description string `json:"description"`

	ConnectionName   string   `json:"connectionName"`
	SpecId           string   `json:"specId"`
	ImageId          string   `json:"imageId"`
	VNetId           string   `json:"vNetId"`
	SubnetId         string   `json:"subnetId"`
	SecurityGroupIds []string `json:"securityGroupIds"`
	SshKeyId         string   `json:"sshKeyId"`
	VmUserAccount    string   `json:"vmUserAccount"`
	VmUserPassword   string   `json:"vmUserPassword"`
}
```

주요 논의 사항: 
- JSON tag 를 삭제하여, 자동으로 변환되게 할 것인지? 
- 변수명과 JSON 필드의 명칭을 완전히 동일하게 처리하도록 할 것인지?

**Comment by jihoon-seo** [comment]
저의 선호는 다음과 같습니다 😊

1. 네이밍 컨벤션
- camelCase로 통일하는 것 좋습니다.
- "의도하지 않은 오픈" 관련: 
  - 저는 '의도하지 않은 오픈이 발생하더라도 첫 글자를 대문자로 처리' 하는 것을 선호합니다.
  - 현황: 현재 거의 모든 TB struct 및 하위 필드가 대문자로 시작하여 export 되어 있는 것으로 보입니다.
- 다음 스프레드시트에, 현황 및 저의 제안을 적어 보겠습니다.
  - https://docs.google.com/spreadsheets/d/1Gu24UN7h2VNBDZGdevxLv54Gr6h3Va4rhuuBCkOvlUw/edit?usp=sharing
2. Golang struct JSON Tag
- JSON tag 를 삭제하는 쪽을 선호합니다.
  - 이렇게 하면, 변수명과 JSON 필드의 명칭을 완전히 동일하게 처리하는 효과가 날 것으로 보입니다. 😊

---

For anyone's information: CB-Tumblebug 팀에서 2019년 9월 17일에 검토 후 작성한 자료
(MCIR stuct들의 field name의 선정/결정 과정 및 초기 버전을 볼 수 있습니다)
https://docs.google.com/presentation/d/1gj-Vlhmxzf1hc8WZvbX8C4gYWbu7Ql6Aj_bjgd9iPU4/edit?usp=sharing
